crates/ruff_python_ast/src/visitor.rs

```diff
@@ -754,3 +797,305 @@ pub fn walk_cmp_op<'a, V: Visitor<'a> + ?Sized>(visitor: &mut V, cmp_op: &'a Cmp
+        let source = "[x for x in numbers]";
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn dict_comprehension() {
+        let source = "{x: x**2 for x in numbers}";
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn set_comprehension() {
+        let source = "{x for x in numbers}";
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn match_class_pattern() {
+        let source = r#"
+match x:
+    case Point2D(0, 0):
+        ...
+    case Point3D(x=0, y=0, z=0):
+        ...
+"#;
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn decorators() {
+        let source = r#"
+@decorator
+def a():
+    pass
+
+@test
+class A:
+    pass
+"#;
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn type_aliases() {
+        let source = r#"type X[T: str, U, *Ts, **P] = list[T]"#;
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn class_type_parameters() {
+        let source = r#"class X[T: str, U, *Ts, **P]: ..."#;
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    #[test]
+    fn function_type_parameters() {
+        let source = r#"def X[T: str, U, *Ts, **P](): ..."#;
+
+        let trace = trace_visitation(source);
+
+        assert_snapshot!(trace);
+    }
+
+    fn trace_visitation(source: &str) -> String {
+        let tokens = lex(source, Mode::Module);
+        let parsed = parse_tokens(tokens, Mode::Module, "test.py").unwrap();
+
+        let mut visitor = RecordVisitor::default();
+        walk_module(&mut visitor, &parsed);
+
+        visitor.output
+    }
+
+    fn walk_module<'a, V>(visitor: &mut V, module: &'a ast::Mod)
+    where
+        V: Visitor<'a> + ?Sized,
+    {
+        match module {
+            ast::Mod::Module(ast::ModModule {
+                body,
+                range: _,
+                type_ignores: _,
+            }) => {
+                visitor.visit_body(body);
+            }
+            ast::Mod::Interactive(ast::ModInteractive { body, range: _ }) => {
+                visitor.visit_body(body);
+            }
+            ast::Mod::Expression(ast::ModExpression { body, range: _ }) => visitor.visit_expr(body),
+            ast::Mod::FunctionType(ast::ModFunctionType {
+                range: _,
+                argtypes,
+                returns,
+            }) => {
+                for arg_type in argtypes {
+                    visitor.visit_expr(arg_type);
+                }
+
+                visitor.visit_expr(returns);
+            }
+        }
+    }
+
+    /// Emits a `tree` with a node for every visited AST node (labelled by the AST node's kind)
+    /// and leafs for attributes.
```

Comment:
Is this identical to the version in `preorder.rs`? Is there any way we can share these utilities?

Reply:
It implements a different `Visitor` trait. But we could implement both `Visitor` traits for `RecordVisitor` and share the `enter_node`, `exit_node`, and `emit` functionality. 

One thing that we could consider adding (out of scope for this PR) is that we have a `pre_visit_node(node: AnyNodeRef)` and maybe `post_visit_node(node: AnyNodeRef)` visitor functions. This would make it easier to implement `Visitor`'s like this (or the comments extraction visitor in the formatter) that want to visit all nodes.

Reply:
The biggest problem here was `visit/walk_module` was missing from `Visitor` but present in `PreorderVisitor`. We can explore a consolidated test pattern in a follow-up. I'll make sure there's a tracking issue before this is merged.